doc/configuration.md

```diff
@@ -0,0 +1,170 @@
+
+## Design
+
+The system supports default configurations with user-specified overrides via configuration files or CLI parameters.
+
+### Key Features
+
+- **Config Validation**: Validates user-provided settings against supported options, rejecting invalid values during deserialization.
+- **Built-in Defaults**: The system defines defaults directly in the code, eliminating the need for a separate default config file.
+- **Merging of Configurations**: Combines configuration sources hierarchically, applying defaults first, followed by file-based overrides, and finally CLI parameters for the highest precedence.
+
+## Config File Structure
+
+**Format**: TOML
+
+- Users can override default settings by supplying a single configuration file or multiple `.toml` files in a directory.
+
+### CLI Parameters
+
+Example: `--config /etc/azure-init/`
+
+## Configuration Hierarchy for `azure-init`
+
+Configuration can be set via a single file, a directory containing multiple files, or default values defined in the code.
+
+### Configuration Loading Order
+
+#### 1. CLI Override (`--config`)
+
+- The `--config` flag specifies the path to load the configuration from, which can be either a single file or a directory.
+  - **File:** If the path points to a file, only that file is loaded.
+    - **Directory:** If the path points to a directory, a series of `.toml` files in the specified directory will be loaded and merged based on specific rules.
+  - **Example:** `azure-init --config /path/to/custom-config.toml`
+
+#### 2. Directory Loading Logic
+
+- If the `--config` parameter points to a directory, `azure-init` loads a base `azure-init.toml` file and merges additional `.toml` files from a `.d` subdirectory in lexicographical order.
+
+     ```text
+     /etc/azure-init/
+     ├── azure-init.toml            # Base configuration
+     └── azure-init.toml.d/
+         ├── 01-network.toml        # Additional network configuration
+         ├── 02-ssh.toml            # Additional SSH configuration
+         └── 99-overrides.toml      # Final overrides
+     ```
+
+#### 3. Defaults in Code
+
+- If no `--config` is specified, the system users built-in defaults.
+
+### Example: Directory with Multiple .toml Files
+
+**Command:**
+
+```sh
+azure-init --config /path/to/custom-config-directory
+```
+
+**Directory Structure:**
+
+```bash
+/path/to/custom-config-directory/
+├── azure-init.toml                # Base configuration
+└── azure-init.toml.d/
+    ├── 01-network.toml            # Network configuration
+    ├── 02-ssh.toml                # SSH configuration
+    └── 99-overrides.toml          # Overrides
+```
+
+**Order of Merging:**
+
+1. Loads `azure-init.toml` as the base configuration.
+2. Merges `.d` files in alphabetical order:
```

Comment:
We mention lexicographical order, but that will only dictate which file is being parses first (you sort them in some order and read them). We need to be explicit about whether the first value sticks, or last value sticks.

Reply:
Last file in the sorted order (in this example, it would be the `99-overrides.toml` file) sticks. I'll make that clearer in the document.